projects/control-service/projects/pipelines_control_service/src/main/java/com/vmware/taurus/service/deploy/DataJobsSynchronizer.java

```diff
@@ -0,0 +1,72 @@
+/*
+ * Copyright 2021-2023 VMware, Inc.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package com.vmware.taurus.service.deploy;
+
+import com.vmware.taurus.service.kubernetes.DataJobsKubernetesService;
+import com.vmware.taurus.service.model.DataJob;
+import com.vmware.taurus.service.model.DataJobDeployment_;
+import com.vmware.taurus.service.model.DataJob_;
+import com.vmware.taurus.service.repository.JobsRepository;
+import io.kubernetes.client.openapi.ApiException;
+import org.apache.commons.lang3.StringUtils;
+import org.springframework.data.domain.Sort;
+import org.springframework.scheduling.concurrent.ThreadPoolTaskExecutor;
+import org.springframework.stereotype.Component;
+
+import java.util.Set;
+
+@Component
+public class DataJobsSynchronizer {
+
+  private final JobsRepository jobsRepository;
+
+  private final DeploymentServiceV2 deploymentService;
+
+  private final DataJobsKubernetesService dataJobsKubernetesService;
+
+  public DataJobsSynchronizer(
+      JobsRepository jobsRepository,
+      DeploymentServiceV2 deploymentService,
+      DataJobsKubernetesService dataJobsKubernetesService) {
+    this.jobsRepository = jobsRepository;
+    this.deploymentService = deploymentService;
+    this.dataJobsKubernetesService = dataJobsKubernetesService;
+  }
+
+  public void synchronizeDataJobs() throws ApiException {
+    ThreadPoolTaskExecutor taskExecutor = initializeTaskExecutor();
+    Set<String> cronJobs = dataJobsKubernetesService.listCronJobs();
+
+    jobsRepository
+        .findAll(
+            Sort.by(
+                Sort.Direction.ASC,
+                DataJob_.DATA_JOB_DEPLOYMENT + "." + DataJobDeployment_.DEPLOYMENT_VERSION_SHA))
+        .forEach(dataJob -> taskExecutor.execute(() -> synchronizeDataJob(dataJob, cronJobs)));
+
+    taskExecutor.shutdown();
+  }
+
+  public void synchronizeDataJob(DataJob dataJob, Set<String> cronJobs) {
+    // Sends notification only when the deployment is initiated by the user.
+    boolean sendNotification =
+        StringUtils.isEmpty(dataJob.getDataJobDeployment().getDeploymentVersionSha());
+    deploymentService.updateDeployment(dataJob, sendNotification, cronJobs);
+  }
+
+  private ThreadPoolTaskExecutor initializeTaskExecutor() {
+    ThreadPoolTaskExecutor executor = new ThreadPoolTaskExecutor();
```

Comment:
I initially thought about this, but I couldn't figure out how to wait until the queue was completely empty.

Reply:
I am thinking here ...

You probably can guess. But it's a bit problematic calling thread pool executor multiple times
There is an  overhead which can become significant if synchronizeDataJobs() is called frequently. Those are resource heavy calls (creating a thread are system calls)

Using CountDownLatch might be an option 

```
 CountDownLatch latch = new CountDownLatch(dataJobs.size());
 synchronizeDataJob(dataJob, cronJobs, latch);  // latch.countDown() called inside 

    latch.await();  // Wait for all tasks to complete
``` 
 
 Though if `synchronizeDataJob` is async annotated method it must be in separate bean or Spring's AOP proxies won't be able to intercept the call. This is something to be aware just in case since it caused me some troubles back in the day until we figure it out.

Reply:
Thanks, I will take it into account.